packages/edit-widgets/src/blocks/legacy-widget/edit/index.js

```diff
@@ -34,71 +39,76 @@ function LegacyWidgetEdit( {
 	const [ isPreview, setIsPreview ] = useState( false );
 	const shouldHidePreview = ! isPreview && hasEditForm;
 
-	function changeWidget() {
-		switchToEdit();
+	const onResetWidget = useCallback( () => {
+		setIsPreview( false );
 
 		setAttributes( {
 			instance: {},
 			id: undefined,
+			referenceWidgetName: undefined,
 			widgetClass: undefined,
 		} );
 		setHasEditForm( true );
-	}
-
-	function switchToEdit() {
-		setIsPreview( false );
-	}
+	} );
 
-	function switchToPreview() {
-		setIsPreview( true );
-	}
+	const visibleLegacyWidgets = useMemo(
+		() => pickBy( availableLegacyWidgets, ( { isHidden } ) => ! isHidden ),
+		[ availableLegacyWidgets ]
+	);
 
 	if ( ! WPWidget ) {
-		return (
-			<LegacyWidgetPlaceholder
-				availableLegacyWidgets={ availableLegacyWidgets }
-				onChangeWidget={ ( newWidget ) => {
-					const {
-						isReferenceWidget,
-						id_base: idBase,
-					} = availableLegacyWidgets[ newWidget ];
+		if ( isEmpty( visibleLegacyWidgets ) ) {
+			return (
+				<Placeholder label={ __( 'Legacy Widget' ) }>
+					{ __( 'There are no widgets available.' ) }
+				</Placeholder>
+			);
+		}
 
-					if ( isReferenceWidget ) {
-						setAttributes( {
-							instance: {},
-							idBase,
-							referenceWidgetName: newWidget,
-							widgetClass: undefined,
-						} );
-					} else {
-						setAttributes( {
-							instance: {},
-							idBase,
-							referenceWidgetName: undefined,
-							widgetClass: newWidget,
-						} );
-					}
-				} }
-			/>
+		return (
+			<Placeholder
+				icon={ <BlockIcon icon={ brush } /> }
+				label={ __( 'Legacy Widget' ) }
+			>
+				<WidgetSelectControl
+					availableLegacyWidgets={ availableLegacyWidgets }
+					onChange={ setAttributes }
+				/>
+			</Placeholder>
 		);
 	}
 
-	const inspectorControls = WPWidget ? (
-		<InspectorControls>
-			<LegacyWidgetInspectorCard
-				name={ WPWidget.name }
-				description={ WPWidget.description }
-			/>
-		</InspectorControls>
-	) : null;
+	if ( isDuplicateReferenceWidget ) {
+		return (
+			<Placeholder
+				icon={ <BlockIcon icon={ brush } /> }
+				label={ __( 'Duplicate Widget' ) }
+			>
+				<p className="components-placeholder__error">
```

Comment:
Hmm, this classname doesn't really seem to do much. I see the styles are defined in the placeholder styles, but I think it must be legacy given the lack of any notable styling. I'd expect some red text or something.

Ideally there would be a `PlaceholderError` component or an `error` prop on the placeholder that renders this. The pattern of using another component's classname like this is discouraged. Some more info:
https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#naming

I've switched this to use the `instructions` prop, which makes it look pretty much the same, but without the code quality issues.